dali/pipeline/operators/reader/loader/sequence_loader.cc

```diff
@@ -0,0 +1,103 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <glob.h>
+#include <algorithm>
+#include <map>
+
+#include "dali/pipeline/operators/reader/loader/sequence_loader.h"
+#include "dali/util/file.h"
+
+namespace dali {
+
+std::string parent_dir(std::string path) {
+  size_t idx = path.rfind("/");
+  DALI_ENFORCE(idx != std::string::npos, "NO PARENT DIRECTORY FOR GIVEN PATH");
+  return path.substr(0, idx + 1);  // return with trailing "/"
+}
+
+void SequenceLoader::PrepareEmpty(TensorSequence *sequence) {
+  sequence->tensors.resize(sequence_length_);
+  for (auto &t : sequence->tensors) {
+    PrepareEmptyTensor(&t);
+  }
+}
+
+void SequenceLoader::ReadSample(TensorSequence *sequence) {
+  // TODO(klecki) this is written as a prototype for video handling
+  const auto &stream = streams_[current_stream_];
+  // TODO(klecki) we probably should buffer the "stream", or recently used
+  // frames
+  for (int i = 0; i < sequence_length_; i++) {
+    LoadFrame(stream, current_frame_ + i, &sequence->tensors[i]);
+  }
+  current_frame_++;
+  // wrap-around
+  if (current_frame_ == stream_sizes_[current_stream_]) {
+    current_stream_++;
+    current_frame_ = 0;
+  }
+  if (current_stream_ == streams_.size()) {
+    current_stream_ = 0;
+  }
+}
+
+Index SequenceLoader::Size() {
+  return total_size_;
+}
+
+std::vector<SequenceLoader::Stream> SequenceLoader::ParseStreams(string file_root) {
+  glob_t glob_buff;
+  std::string glob_pattern = file_root + "/*/*";
+  const int glob_flags = 0;
+  glob(glob_pattern.c_str(), glob_flags, nullptr, &glob_buff);
+  std::vector<std::string> files;
+  for (size_t i = 0; i < glob_buff.gl_pathc; i++) {
+    files.emplace_back(glob_buff.gl_pathv[i]);
+  }
+  std::sort(files.begin(), files.end());
+  std::map<std::string, size_t> streamName_bucket_map;
+  std::vector<Stream> streams;
+  for (auto &f : files) {
+    auto parent = parent_dir(f);
+    auto bucket = streamName_bucket_map.find(parent);
+    if (bucket == streamName_bucket_map.end()) {
+      streams.push_back({parent, {f}});
+      streamName_bucket_map[parent] = streams.size() - 1;
+    } else {
+      streams[bucket->second].second.push_back(f);
+    }
+  }
+  return streams;
+}
+
+vector<size_t> SequenceLoader::CalculateStreamSizes(const vector<Stream> &streams,
+                                                    size_t sample_lenght) {
+  vector<size_t> result;
+  for (auto &s : streams) {
+    result.push_back(s.second.size() - (sample_lenght - 1));
+  }
+  return result;
+}
+
+void SequenceLoader::LoadFrame(const Stream &s, Index frame_idx, Tensor<CPUBackend> *target) {
+  const auto frame_filename = s.second[frame_idx];
+  FileStream *frame = FileStream::Open(frame_filename);
+  Index frame_size = frame->Size();
+  target->Resize({frame_size});
+  frame->Read(target->mutable_data<uint8_t>(), frame_size);
+  frame->Close();
```

Comment:
I think I'm lost, but where is the implementation of `Close` and `Read`? Looks like you are creating `FileStream` object in line 96. When I got to `file.h` I see:
```
  static FileStream * Open(const std::string& uri);

  virtual void Close() = 0;
  virtual size_t Read(uint8_t * buffer, size_t n_bytes) = 0;
```

Reply:
The `Open` is actually a factory-like method, that returns pointer to a subclass `LocalFileStream`, that has all those methods implemented, for details see dali\util\local_file.cc

Reply:
In my opinion, this design is extremely erroneous and should be changed as soon as possible. E.g. in this PR :)

Reply:
We should keep separate things in separate PRs.

Reply:
You still need to delete this FileStream object. #308 will address this, but until then please fix it here.

Reply:
Done

Reply:
You need to free `frame`.

Reply:
I put in unique_ptr, so it will be freed after it goes out of scope.